### ISSUE TITLE
Proposal: Grammatical Typology to Enable Advanced Tooling

### ISSUE THREAD

**Issue by aronallen** [mostjs/core#210]
Crosspost from https://github.com/cujojs/most/issues/516

Observables are one of the most general abstractions we have available in Computer Science. They provide a common interface to operate on singular or plural and synchronous or asynchronous items.

But programming with Observables is a path riddled with traps, partly due to its novelty, but also that our tooling mostly is built around synchronous imperative code, and not asynchronous functional code.

Typing can get us really far in removing common errors when piping Observables through our code, but our types are not going to help us with the problem of time.

I have spent some days philosophizing whether we need a temporal-type to decorate our Observables, to get a better immediate idea of the consequences when merging, combining, delaying, debouncing, skipping, etc...

At first I set out to categories Observables in my own made-up categories, but then I realized we already have a quite extensive and rigid categorization of temporal events, this system is known as Grammar more properly conjugation, and it provides us with an opportunity to type our Observables not just with their expected value, but when that value might arrive.

In Latin conjugation we can operate with four dimensions, mood, voice, tense, and person. These four dimensions map well to what we need to properly time-categorize Streams.

### Grammatical Mood

Grammatical mood helps us know the certainty of an event. Moods can be divided into Realis and Irrealis, things that are certain, and things that are uncurtain. Indicative is a Realis mood, the rest are Irrealis.

They may not all be applicable for Observables

```
enum Mood {
  Indicative, // realis certain
  Subjunctive, // it might
  Conditional, // if
  Optative, // hope
  Imperative, // command
  Jussive, // command to non-present
  Potential, // it could
  Hypothetical, // it is unlikely
  Inferential, // tree falling in forest without observer
  Interrogative // asking (probably better for a pull based thing)
}
```

### Grammatical Voice

Voice indicates whether something is doing something, or something is happening to something. In observables it could indicate whether the Observable producer has all it's futures defined, or whether it awaits input from the outside (such as passing of time, mouse clicks).
Most Observables would be Passive, with the exception of Observables created from arrays or constants.
```

enum Voice {
  Acitve,
  Passive
}
```

### Grammatical Tense

Grammatical tense indicates when something may happen, or when something did happen, we might want to add a Never, or Never may simply be an undefined Tense.

```
enum Tense {
  Present, // now
  Imperfect, // past and ongoing
  Perfect, //past and complete
  Future, // soon
  Pluperfect, // happened before something else in the past
  FuturePerfect // will happen at this time
};
```

### Gramatical Person

Finally we have grammatical persons, that indicates who is doing what. We could use the person to indicate singular and plural events of a stream, and the degree of stream (higher-order-streams)

```
enum Person {
  FirstPersonSingular, // a stream of one thing
  FirstPersonPlural, // a stream of many things
  SecondPersonSinglar, // a stream of one stream of one thing
  SecondPersonPlural, // a stream of streams of things.
  ThirdPersonSingular, // a stream of one stream of one stream of one thing
  ThirdPersonPlural // a stream of streams of streams of things
}
```

Once our Observable producers are properly decorated with grammatical metadata. We can proceed to ensure that our operators determine what the conjugation of the resulting stream would be.

### Practical Applications

Imagine you are combining stream a$ and b$, they are both ```Stream<number>``` but a$ is defined far away from where you are currently working, a$ happens to be ```never()``` while b$ is ```just(3)``` but you are unaware and spend some time trying to find the culprit. With gramatical typings you could get a warning, indicating you are trying to combine something b$ (Present, FirstPersonSingular, Active, Indicative) with a$ (Never), you would get a hint as to where something is amiss.

I would welcome some input on the ideas above, is it useful?

**Comment by aronallen** [comment]
I posted this here to continue the discussion. 

Another approach would be to reuse the grammatical dimensions, but instead of using linguistic terms, we map everything to numbers.

Tense: integer, a negative number could indicate that events happened in the past, but no more values are expected. A positive number could indicate that events will happen in the future, and 0 would indicate that events will happen upon subscription whereafter it will synchronously complete. NaN could represent never(). just(0).delay(1000) would result in a stream with a tense of 1000

Mood: could be a float, indicating the certainty of the past, present, or future. 0 indicates that we are completely certain nothing happened within the tense, 1 indicates that we are absolutely sure something happened, is happening, or will happen. The certainty could be anything between 0 and 1.

Number: integer, the number of events in the stream in the best case, could be 0, 1, 3, Infinity.

Passive: integer, wether the stream producer is waiting for outside effects (clicks, passing of time), or whether it produces on it's own ```fromArray([1,2,3])```

Order: the degree of the stream.

We should also consider if we want to grammatically describe the Observable as a whole, or if we want to assign a grammatical type to each channel (next, error, complete).

Anyway working with numbers instead of categories could simplify the logic when determining the resulting stream, the grammatical type could be stored as an array of 5 numbers.

**Comment by aronallen** [comment]
I also put a draft summary on the TypeScript repo.
I imagine Conjugation/Inflection could have many applications beyond Observables.
https://github.com/Microsoft/TypeScript/issues/21424

**Comment by Frikki** [comment]
I like the idea. However, I am worried that this introduces a complexity that makes it hard for people to use this otherwise relatively simple library. In other words, Latin grammar is not commonly understood.

**Comment by aluanhaddad** [comment]
I think the idea of applying grammatical concepts to better understand our programs is fascinating.

 That said, the statement that

> Observables are one of the most general abstractions we have available in Computer Science. They provide a common interface to operate on singular or plural and synchronous or asynchronous items

while technically true, troubles me.

Erik Meijer has stated a number of times that observables were created to and designed for handling the asynchronous plural case.

Trying to replace all other programming constructs with observables just because they can represent them strikes me as a bad idea.


In other words
```ts
Promise<string> retrieveSomething()
```
is clear and better than
```ts
FirstPersonSingular Future Observable<string> retrieveSomething()
```

**Comment by psirenny** [comment]
> I also put a draft summary on the TypeScript repo.

Yea, this does strike me as something that would need to be enforced by the type system for it to have tangible benefits.

It would be useful to see lots of examples of things named using the latin concepts that you've described. If anything, it could help in the naming of observable variables in user code.

Naming things is something that we all constantly struggle with and having a naming convention would be helpful. I think the subject would make a good blog post and be relevant to other forms of asynchronous programming.

One thing I will say, is that I think the language constructs will need to be in English for it to gain traction.

**Comment by briancavalier** [comment]
@aronallen Thanks for posting these interesting ideas.  I'm definitely in favor of finding ways types and type-level logic can help.

We've talked a bit about how we could use types to provide more information about most streams.  Some of them are akin to what's possible in dependently typed languages.  For example, Flow has type level functions (I'm more familiar with Flow than TS), and so type-level computations are possible.  

> With gramatical typings you could get a warning

Yes. I think the goal of these kinds of types should to prevent you ever constructing nonsensical stream combinations in the first place.

I think some notion of proportion (number of events) would be useful, though I'm not sure what representation or how fined grained would turn out to be most helpful.  For example, I can imagine zero, one, zero or more, and one or more being useful.

Active/Passive seems potentially useful, as does Order.  Although, it seems like Order is already represented in the type: In `Stream<A>`, `A` could be `Stream<number>` which indicates a second order stream.  The singular vs plural seems like it would be represented in the proportion.  Am I thinking about that correctly, or is there additional information that can be expressed by having specific Order types?

Mood and Tense are harder for me to envision in practical use.  Have you thought of a simple concrete example that might help me get my head around how this could help in practice?

@aluanhaddad

> Trying to replace all other programming constructs with observables just because they can represent them strikes me as a bad idea.

I agree with this.  I've noticed that when some folks first encounter most streams, they try to use them for _everything_.  I usually try to nudge those folks gently to use the simplest structure that is appropriate for each use case.

It's useful in the context of programming with Streams to have a way to construct a stream containing a single event at a particular time, just as it's useful to be able to construct a singleton list in the context of programming with lists.  However, because it's possible to construct an errored stream and a stream containing a single event doesn't mean that streams should be used in place of a simpler Either.

**Comment by aronallen** [comment]
Thanks for all the feedback. I really appreciate it, I think the idea I proposed is a bit rough, and needs to be refined by many bright minds.

I will try to get back on most of the concerns and questions raised.

First off I would like to clarify that I don't believe every problem in computer science should be solved with the Observable pattern. There are some problems that are very synchronous, or very pull based in nature, that simply don't map well to Observables. Though specifically I still find singular Observables superior to Promises, mainly because Observables can support cancelation logic.

Secondly I do not want Inflections to make Observables in general, or Most in particular more difficult to learn or use. Actually I think these extensions could help developers learning Observables to get a better immediate idea of what is going on.

Thirdly the reason I took outset in the grammar-system provided by Latin, is because it is quite extensive. If we are to develop this idea we need to find a language that strikes a balance between familiarity and accuracy. Whatever this idea develops into, it should be useful to people who hold no undergraduate degree in latin or computer science. I have no higher education so this is of utmost importance for me. Plotting things visually and explaining them on a time-line in a editor extension could also help developers understand what the inflections actually mean.

I don't imagine that developers will need to diligently Inflect their streams in correct tense, mood, voice, number, etc. Rather the library should do the grammatical heavy lifting, so developers can mostly rely on type inference. I also imagine that Inflection types should not raise errors and only raise warnings, as the developer may have a good reason to create an observable that never resolves.

With very complex chains of combination, merge, delay, it may be hard to infer any correct inflection, in such a case the result would still be Observable of something, and should still be able to be used.

@briancavalier you bring some great points, the dimension of order is a bit irrelevant as this is already covered by our existing system, and this is a problem of value-types, not time-types.

I will try to provide some examples on how mood and tense specifically can be useful.

```typescript
const click$: Inflection<Observable<Event>, Future, Subjunctive, Passive, Plural>
 = fromEvent('click', element);
```

The click event stream is Future, because there are known to be no events in the present, Subjunctive, because it is a possible future, but not a definite future. Passive because it is acted upon (by clicking the mouse) and plural because there could be more than one event.

```
// we take the click event from above and extract the clientX coordinate
// it retains the grammatical properties of click$ 
const x$ = click$.map(event => event.clientX);
// we define a new stream of just a number
const y$: Inflection<Observable<number>, Present, Indicative, Active, Singular> = just(1);
// now we operate on the streams from above
// in the case of combine, the resulting inflection, would be the latest, least likely, and minimum case
// thus our typesystem should infer that xy$ would be Future, Subjunctive, Passive, Singular.
const xy$ = combine((a, b) => a * b, x$, y$);
// in the case of merge, our type system should infer the earliest, most likely, most numerous case
// in this case it would be PresentContinous, Indicative, Active, Plural. 
const yx$ = merge(x$, y$)
```

from the example above it may be hard to decide when things are active or passive, we need a convention here, which could be that the output is only active, if all inputs are active.

Specifically mood can help us speak of the certainty of the events. Indicative is certain, whereas all other moods express possibility, opportunity or likelihood.

Tense can help us know if something is already complete.
Imagine the following case, I will omit the Observable decorations for now, and just focus on the Inflection.

```typescript

const a$: Inflection<Present, Indicative, Active, Singular> = just(1);
const b$: Inflection<Future, Indicative, Passive, Singular> = delay(100, $a);
const c$: Inflection<Perfect, Indicative, Active, Nil> = a$.skip(1);

```

**Comment by aronallen** [comment]
I have given this more thought. In order for this to work, we at least need to basic operators for combination and merge. 
We could define them as AND and OR.

The OR operator would result in the earliest tense, the most real mood, the most quntative number.

The AND operator would result in the latest tense, the most irreal mood, and the least quantitative number. 

For both operators voice will only be active if all inputs are active.

With these two primitive operators it should be possible to express most operators.

For schedulers we need something different.

**Comment by Frikki** [comment]
Since the discussion seems to have died off, I am closing the issue.
